## Работа с данными в Pandas

In [1]:
import numpy as np
import pandas as pd

pd.set_option("display.precision", 2)

In [2]:
DATA_PATH = "telecom_churn.csv"

In [ ]:
"filename.parquet"

### Основные объекты Pandas

В библиотеке pandas определены два класса объектов для работы с данными:

**Series** — одномерный массив, который может хранить значения любого типа данных;

**DataFrame** — двумерный массив (таблица), в котором столбцами являются объекты класса Series.


### Основные методы Pandas

Прочитаем данные (используя метод read_csv) и посмотрим на первые 5 строк, используя метод head:

In [3]:
df = pd.read_csv(DATA_PATH)
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

In [13]:
df["State"]

0       KS
1       OH
2       NJ
3       OH
4       OK
        ..
3328    AZ
3329    WV
3330    RI
3331    CT
3332    TN
Name: State, Length: 3333, dtype: object

Посмотрим на ее размерность

In [4]:
print(df.shape)

(3333, 20)


Видно, что мы имеем дело с 20 столбцами и 3333 строками

Теперь посмотрим на названия столбцов в датасете

In [6]:
print(list(df.columns))

['State', 'Account length', 'Area code', 'International plan', 'Voice mail plan', 'Number vmail messages', 'Total day minutes', 'Total day calls', 'Total day charge', 'Total eve minutes', 'Total eve calls', 'Total eve charge', 'Total night minutes', 'Total night calls', 'Total night charge', 'Total intl minutes', 'Total intl calls', 'Total intl charge', 'Customer service calls', 'Churn']


Выведем общую информацию о столбцах.

К общей информации относится: индекс, название столбцов, кол-во не нулевых элементов в наборе и тип данных в столбце.

Важно отметить, что елси тип данных во всем столбце невозможно однозначно определить, то столбцу назначается тип Object.

In [7]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

Теперь давайте преобразуем тип в рамках столбца. К примеру из bool переведем в int64.

In [8]:
df["Churn"] = df["Churn"].astype("int64")

Проверим этот факт в информации о таблице.

In [9]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

#### Таким нехитрым способом можно вывести общие статистики вещественных фичей по столбцам во всей таблице

In [10]:
df.describe()

Account length  Area code  Number vmail messages  Total day minutes  \
count         3333.00    3333.00                3333.00            3333.00   
mean           101.06     437.18                   8.10             179.78   
std             39.82      42.37                  13.69              54.47   
min              1.00     408.00                   0.00               0.00   
25%             74.00     408.00                   0.00             143.70   
50%            101.00     415.00                   0.00             179.40   
75%            127.00     510.00                  20.00             216.40   
max            243.00     510.00                  51.00             350.80   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count          3333.00           3333.00            3333.00          3333.00   
mean            100.44             30.56             200.98           100.11   
std              20.07              9.26              50.71            19.92   
min               0.00              0.00               0.00             0.00   
25%              87.00             24.43             166.60            87.00   
50%             101.00             30.50             201.40           100.00   
75%             114.00             36.79             235.30           114.00   
max             165.00             59.64             363.70           170.00   

       Total eve charge  Total night minutes  Total night calls  \
count           3333.00              3333.00            3333.00   
mean              17.08               200.87             100.11   
std                4.31                50.57              19.57   
min                0.00                23.20              33.00   
25%               14.16               167.00              87.00   
50%               17.12               201.20             100.00   
75%               20.00               235.30             113.00   
max               30.91               395.00             175.00   

       Total night charge  Total intl minutes  Total intl calls  \
count             3333.00             3333.00           3333.00   
mean                 9.04               10.24              4.48   
std                  2.28                2.79              2.46   
min                  1.04                0.00              0.00   
25%                  7.52                8.50              3.00   
50%                  9.05               10.30              4.00   
75%                 10.59               12.10              6.00   
max                 17.77               20.00             20.00   

       Total intl charge  Customer service calls    Churn  
count            3333.00                 3333.00  3333.00  
mean                2.76                    1.56     0.14  
std                 0.75                    1.32     0.35  
min                 0.00                    0.00     0.00  
25%                 2.30                    1.00     0.00  
50%                 2.78                    1.00     0.00  
75%                 3.27                    2.00     0.00  
max                 5.40                    9.00     1.00

Также можно посмотреть разные статистики отдельно по столбцам

In [11]:
df["Churn"].mean(), df["Churn"].min(), df["Churn"].max(), df["Churn"].std(), df["Churn"].median()

(0.14491449144914492, 0, 1, 0.35206742362412946, 0.0)

Если задать уточнение типов, которые нас интересуют, то можно собрать более узкий срез статистик

In [14]:
df.describe(include=["object", "bool"])

State International plan Voice mail plan
count   3333               3333            3333
unique    51                  2               2
top       WV                 No              No
freq     106               3010            2411

Также можем посмотреть на количество уникальных элементов в столбце(гистограмма)

In [ ]:
df["Churn"].value_counts()

Churn
0    2850
1     483
Name: count, dtype: int64

Дополнительным параметром `normalize=True` можно нормализовать переменные.

In [15]:
df["Churn"].value_counts(normalize=True)

Churn
0    0.86
1    0.14
Name: proportion, dtype: float64

### Сортировка

Отсортируем таблицу по столбцу `Total day charge` в порядке невозрастания.

In [16]:
df.sort_values(by="Total day charge", ascending=False).head()

State  Account length  Area code International plan Voice mail plan  \
365     CO             154        415                 No              No   
985     NY              64        415                Yes              No   
2594    OH             115        510                Yes              No   
156     OH              83        415                 No              No   
605     MO             112        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
365                       0              350.8               75   
985                       0              346.8               55   
2594                      0              345.3               81   
156                       0              337.4              120   
605                       0              335.5               77   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
365              59.64              216.5               94             18.40   
985              58.96              249.5               79             21.21   
2594             58.70              203.4              106             17.29   
156              57.36              227.4              116             19.33   
605              57.04              212.5              109             18.06   

      Total night minutes  Total night calls  Total night charge  \
365                 253.9                100               11.43   
985                 275.4                102               12.39   
2594                217.5                107                9.79   
156                 153.9                114                6.93   
605                 265.0                132               11.93   

      Total intl minutes  Total intl calls  Total intl charge  \
365                 10.1                 9               2.73   
985                 13.3                 9               3.59   
2594                11.8                 8               3.19   
156                 15.8                 7               4.27   
605                 12.7                 8               3.43   

      Customer service calls  Churn  
365                        1      1  
985                        1      1  
2594                       1      1  
156                        0      1  
605                        2      1

Также можно отсортировать таблицу по нескольким столбцам в определенном порядке для каждого. Для этого необходимо подать два списка в порядках приоритета. В данном случае сначала происходит сортировка по `Churn` в порядке возрастания, а в рамках одинаковых элементов по `Churn` происходит сортировка по `Total day charge` в порядке убывания.

In [17]:
df.sort_values(by=["Churn", "Total day charge"], ascending=[True, False]).head()

State  Account length  Area code International plan Voice mail plan  \
688     MN              13        510                 No             Yes   
2259    NC             210        415                 No             Yes   
534     LA              67        510                 No              No   
575     SD             114        415                 No             Yes   
2858    AL             141        510                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
688                      21              315.6              105   
2259                     31              313.8               87   
534                       0              310.4               97   
575                      36              309.9               90   
2858                     28              308.0              123   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
688              53.65              208.9               71             17.76   
2259             53.35              147.7              103             12.55   
534              52.77               66.5              123              5.65   
575              52.68              200.3               89             17.03   
2858             52.36              247.8              128             21.06   

      Total night minutes  Total night calls  Total night charge  \
688                 260.1                123               11.70   
2259                192.7                 97                8.67   
534                 246.5                 99               11.09   
575                 183.5                105                8.26   
2858                152.9                103                6.88   

      Total intl minutes  Total intl calls  Total intl charge  \
688                 12.1                 3               3.27   
2259                10.1                 7               2.73   
534                  9.2                10               2.48   
575                 14.2                 2               3.83   
2858                 7.4                 3               2.00   

      Customer service calls  Churn  
688                        3      0  
2259                       3      0  
534                        4      0  
575                        1      0  
2858                       1      0

In [ ]:
df = df.sort_values(by=["Churn", "Total day charge"], ascending=[True, False])

### Рассмотрим примеры

#### Каковы средние значения числовых характеристик для непостоянных пользователей?

Операция маскирования - очень удобный инструмент когда нам нужно отфильтровать данные по определенному критерию. В данном случае `df["Churn"] == 1` - маска по которой будет фильтроваться набор `df`.

In [18]:
df["Churn"] == 1

0       False
1       False
2       False
3       False
4       False
        ...  
3328    False
3329    False
3330    False
3331    False
3332    False
Name: Churn, Length: 3333, dtype: bool

In [22]:
df[(df["Churn"] == 1) | (df["State"] == "NV")].head(5)

State  Account length  Area code International plan Voice mail plan  \
10    IN              65        415                 No              No   
15    NY             161        415                 No              No   
21    CO              77        408                 No              No   
33    AZ              12        408                 No              No   
41    MD             135        408                Yes             Yes   

    Number vmail messages  Total day minutes  Total day calls  \
10                      0              129.1              137   
15                      0              332.9               67   
21                      0               62.4               89   
33                      0              249.6              118   
41                     41              173.1               85   

    Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
10             21.95              228.5               83             19.42   
15             56.59              317.8               97             27.01   
21             10.61              169.9              121             14.44   
33             42.43              252.4              119             21.45   
41             29.43              203.9              107             17.33   

    Total night minutes  Total night calls  Total night charge  \
10                208.8                111                9.40   
15                160.6                128                7.23   
21                209.6                 64                9.43   
33                280.2                 90               12.61   
41                122.2                 78                5.50   

    Total intl minutes  Total intl calls  Total intl charge  \
10                12.7                 6               3.43   
15                 5.4                 9               1.46   
21                 5.7                 6               1.54   
33                11.8                 3               3.19   
41                14.6                15               3.94   

    Customer service calls  Churn  
10                       4      1  
15                       4      1  
21                       5      1  
33                       1      1  
41                       0      1

`select_dtypes` - позволяет выбрать из набора данных столбцы с определенным типом данных

In [23]:
df.select_dtypes(include=np.number)[df["Churn"] == 1]

Account length  Area code  Number vmail messages  Total day minutes  \
10                65        415                      0              129.1   
15               161        415                      0              332.9   
21                77        408                      0               62.4   
33                12        408                      0              249.6   
41               135        408                     41              173.1   
...              ...        ...                    ...                ...   
3301              84        415                      0              280.0   
3304              71        510                      0              186.1   
3320             122        510                      0              140.0   
3322              62        408                      0              321.1   
3323             117        415                      0              118.4   

      Total day calls  Total day charge  Total eve minutes  Total eve calls  \
10                137             21.95              228.5               83   
15                 67             56.59              317.8               97   
21                 89             10.61              169.9              121   
33                118             42.43              252.4              119   
41                 85             29.43              203.9              107   
...               ...               ...                ...              ...   
3301              113             47.60              202.2               90   
3304              114             31.64              198.6              140   
3320              101             23.80              196.4               77   
3322              105             54.59              265.5              122   
3323              126             20.13              249.3               97   

      Total eve charge  Total night minutes  Total night calls  \
10               19.42                208.8                111   
15               27.01                160.6                128   
21               14.44                209.6                 64   
33               21.45                280.2                 90   
41               17.33                122.2                 78   
...                ...                  ...                ...   
3301             17.19                156.8                103   
3304             16.88                206.5                 80   
3320             16.69                120.1                133   
3322             22.57                180.5                 72   
3323             21.19                227.0                 56   

      Total night charge  Total intl minutes  Total intl calls  \
10                  9.40                12.7                 6   
15                  7.23                 5.4                 9   
21                  9.43                 5.7                 6   
33                 12.61                11.8                 3   
41                  5.50                14.6                15   
...                  ...                 ...               ...   
3301                7.06                10.4                 4   
3304                9.29                13.8                 5   
3320                5.40                 9.7                 4   
3322                8.12                11.5                 2   
3323               10.22                13.6                 3   

      Total intl charge  Customer service calls  Churn  
10                 3.43                       4      1  
15                 1.46                       4      1  
21                 1.54                       5      1  
33                 3.19                       1      1  
41                 3.94                       0      1  
...                 ...                     ...    ...  
3301               2.81                       0      1  
3304               3.73                       4      1  
3320               2.62          

#### Сколько времени (в среднем) непостоянные пользователи проводят на телефоне в дневное время?

Опять же применим операцию маскирования чтобы выделить непостоянных пользователей, а затем выберем нужный нам столбец и посчитаем среднее

In [ ]:
df[df["Churn"] == 1]["Total day minutes"].mean()

206.91407867494823

#### Какова максимальная продолжительность международных звонков среди постоянных пользователей (`Churn == 0`), у которых нет международного тарифного плана?

Здесь нам нужна фильтрация по двум критериям: `Churn == 0` - постоянный пользователь и `International plan == No` - отсутствие международного плана.

In [24]:
df[(df["Churn"] == 0) & (df["International plan"] == "No")]["Total intl minutes"].max()

18.9

### Индексация

Набор данных может быть проиндексирован как по строкам, так и по столбцам. Для этого используется метод `.loc(row range, col range)`

In [25]:
df.loc[0:5, "State":"Area code"]

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415
5    AL             118        510

In [27]:
df.iloc[0:5, 0:3]

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415

Также совершенно спокойно можно выполнять простые срезы по строкам не обращаясь к методу `.loc`

In [28]:
df[-1:]

State  Account length  Area code International plan Voice mail plan  \
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3332             39.85              265.9               82              22.6   

      Total night minutes  Total night calls  Total night charge  \
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
3332                13.7                 4                3.7   

      Customer service calls  Churn  
3332                       0      0

Также очень просто можно применить к набору данных преобразование в виде функции одного аргумента(так называемый Mapping). Функция должна иметь 1 входной аргумент(текущее значение) и 1 выходной аргумент(результат обработки)

In [30]:
df.apply(np.min)

State                       AK
Account length               1
Area code                  408
International plan          No
Voice mail plan             No
Number vmail messages        0
Total day minutes          0.0
Total day calls              0
Total day charge           0.0
Total eve minutes          0.0
Total eve calls              0
Total eve charge           0.0
Total night minutes       23.2
Total night calls           33
Total night charge        1.04
Total intl minutes         0.0
Total intl calls             0
Total intl charge          0.0
Customer service calls       0
Churn                        0
dtype: object

In [31]:
df.head(5)

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

In [32]:
def transform(x):
    if x.dtype == np.number:
        return x * x
    return x

df.apply(transform).head(5)

C:\Users\Kirill\AppData\Local\Temp\ipykernel_11136\3497662165.py:2: DeprecationWarning: Converting `np.inexact` or `np.floating` to a dtype is deprecated. The current result is `float64` which is not strictly correct.
  if x.dtype == np.number:


State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25           70278.01              110   
1                     26           26114.56              123   
2                      0           59243.56              114   
3                      0           89640.36               71   
4                      0           27788.89              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0           2031.30           38966.76               99            281.57   
1            754.60           38220.25              103            276.22   
2           1712.30           14689.44              110            106.09   
3           2590.81            3831.61               88             27.67   
4            803.16           21992.89              122            159.01   

   Total night minutes  Total night calls  Total night charge  \
0             59878.09                 91              121.22   
1             64719.36                103              131.10   
2             26438.76                104               53.58   
3             38769.61                 89               78.50   
4             34931.61                121               70.73   

   Total intl minutes  Total intl calls  Total intl charge  \
0              100.00                 3               7.29   
1              187.69                 3              13.69   
2              148.84                 5              10.82   
3               43.56                 7               3.17   
4              102.01                 3               7.45   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

Также полученные результаты можно использовать для маскирования

In [33]:
df[df["State"].apply(lambda state: state[0] == "W")].head()

State  Account length  Area code International plan Voice mail plan  \
9     WV             141        415                Yes             Yes   
26    WY              57        408                 No             Yes   
44    WI              64        510                 No              No   
49    WY              97        415                 No             Yes   
54    WY              87        415                 No              No   

    Number vmail messages  Total day minutes  Total day calls  \
9                      37              258.6               84   
26                     39              213.0              115   
44                      0              154.0               67   
49                     24              133.2              135   
54                      0              151.0               83   

    Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
9              43.96              222.0              111             18.87   
26             36.21              191.1              112             16.24   
44             26.18              225.8              118             19.19   
49             22.64              217.2               58             18.46   
54             25.67              219.7              116             18.67   

    Total night minutes  Total night calls  Total night charge  \
9                 326.4                 97               14.69   
26                182.7                115                8.22   
44                265.3                 86               11.94   
49                 70.6                 79                3.18   
54                203.9                127                9.18   

    Total intl minutes  Total intl calls  Total intl charge  \
9                 11.2                 5               3.02   
26                 9.5                 3               2.57   
44                 3.5                 3               0.95   
49                11.0                 3               2.97   
54                 9.7                 3               2.62   

    Customer service calls  Churn  
9                        0      0  
26                       0      0  
44                       1      0  
49                       1      0  
54                       5      1

Если мы хотим применить некоторую таблицу в качестве преобразования значений, то можно использовать операцию `.map`

В данном случае все строки `"No"` заменятся на `False`, а `"Yes"` на `True`

In [34]:
d = {"No": False, "Yes": True}
df["International plan"] = df["International plan"].map(d)
df.head()

State  Account length  Area code  International plan Voice mail plan  \
0    KS             128        415               False             Yes   
1    OH             107        415               False             Yes   
2    NJ             137        415               False              No   
3    OH              84        408                True              No   
4    OK              75        415                True              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

Также мы можем применить более специализированную операцию `replace` для того же преобразования

In [36]:
df = df.replace({"Voice mail plan": d, "NAME": d})
df.head()

State  Account length  Area code  International plan  Voice mail plan  \
0    KS             128        415               False             True   
1    OH             107        415               False             True   
2    NJ             137        415               False            False   
3    OH              84        408                True            False   
4    OK              75        415                True            False   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

### Группировка

Давайте посмотрим на различие по статистикам между постоянными и не постоянными пользователями

Для этого нужно применить группировку данных по столбцу `Churn` и посмотреть на определенные столбцы

In [37]:
columns_to_show = ["Total day minutes", "Total eve minutes", "Total night minutes"]

#  Группировка        Выбор столбцов   Показываем статистики(без перцентилей для упрощения)
df.groupby(["Churn"])[columns_to_show].describe(percentiles=[])

Total day minutes                                   Total eve minutes  \
                  count    mean    std  min    50%    max             count   
Churn                                                                         
0                2850.0  175.18  50.18  0.0  177.2  315.6            2850.0   
1                 483.0  206.91  69.00  0.0  217.6  350.8             483.0   

                                         Total night minutes                 \
         mean    std   min    50%    max               count    mean    std   
Churn                                                                         
0      199.04  50.29   0.0  199.6  361.8              2850.0  200.13  51.11   
1      212.41  51.73  70.9  211.3  363.7               483.0  205.23  47.13   

                            
        min     50%    max  
Churn                       
0      23.2  200.25  395.0  
1      47.4  204.80  354.9

In [39]:
df.groupby(by=["Churn"]).agg()

Также мы можем указать аггрегационные функции которые мы хотим видеть

In [40]:
columns_to_show = ["Total day minutes", "Total eve minutes", "Total night minutes"]

df.groupby(["Churn"])[columns_to_show].agg(["mean", "std", "min", "max"])

Total day minutes                    Total eve minutes               \
                   mean    std  min    max              mean    std   min   
Churn                                                                       
0                175.18  50.18  0.0  315.6            199.04  50.29   0.0   
1                206.91  69.00  0.0  350.8            212.41  51.73  70.9   

             Total night minutes                      
         max                mean    std   min    max  
Churn                                                 
0      361.8              200.13  51.11  23.2  395.0  
1      363.7              205.23  47.13  47.4  354.9

Еще одна немаловажная операция - сведение таблиц. Сводные таблицы позволяют выявить связи между двумя или более категориальными переменными.

Таким образом мы можем посмотреть количество постоянных и не постоянных пользователей с международными планами или без них.


In [41]:
pd.crosstab(df["Churn"], df["International plan"])

International plan  False  True 
Churn                           
0                    2664    186
1                     346    137

In [ ]:
pd.merge()

In [42]:
start = df.head()
end = df.tail()

In [43]:
pd.concat([start, end], )

State  Account length  Area code  International plan  Voice mail plan  \
0       KS             128        415               False             True   
1       OH             107        415               False             True   
2       NJ             137        415               False            False   
3       OH              84        408                True            False   
4       OK              75        415                True            False   
3328    AZ             192        415               False             True   
3329    WV              68        415               False            False   
3330    RI              28        510               False            False   
3331    CT             184        510                True            False   
3332    TN              74        415               False             True   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1              110   
1                        26              161.6              123   
2                         0              243.4              114   
3                         0              299.4               71   
4                         0              166.7              113   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4               99             16.78   
1                27.47              195.5              103             16.62   
2                41.38              121.2              110             10.30   
3                50.90               61.9               88              5.26   
4                28.34              148.3              122             12.61   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7                 91               11.01   
1                   254.4                103               11.45   
2                   162.6                104                7.32   
3                   196.9                 89                8.86   
4                   186.9                121                8.41   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
3328                 9.9                 6               2.67   
3329                 9.6                 4               2.59   
3330                14.1                 6               3.81   
3331                 5.0                10               1.35   
3332                13.7                 4               3.70   

      Customer service calls  Churn  
0                       

In [1]:
left = df.loc[:, "State":"Area code"]

NameError: name 'df' is not defined

In [49]:
right = df.loc[:, "Voice mail plan":"Total eve minutes"]

In [50]:
left

State  Account length  Area code
0       KS             128        415
1       OH             107        415
2       NJ             137        415
3       OH              84        408
4       OK              75        415
...    ...             ...        ...
3328    AZ             192        415
3329    WV              68        415
3330    RI              28        510
3331    CT             184        510
3332    TN              74        415

[3333 rows x 3 columns]

In [51]:
right

Voice mail plan  Number vmail messages  Total day minutes  \
0                True                     25              265.1   
1                True                     26              161.6   
2               False                      0              243.4   
3               False                      0              299.4   
4               False                      0              166.7   
...               ...                    ...                ...   
3328             True                     36              156.2   
3329            False                      0              231.1   
3330            False                      0              180.8   
3331            False                      0              213.8   
3332             True                     25              234.4   

      Total day calls  Total day charge  Total eve minutes  
0                 110             45.07              197.4  
1                 123             27.47              195.5  
2                 114             41.38              121.2  
3                  71             50.90               61.9  
4                 113             28.34              148.3  
...               ...               ...                ...  
3328               77             26.55              215.5  
3329               57             39.29              153.4  
3330              109             30.74              288.8  
3331              105             36.35              159.6  
3332              113             39.85              265.9  

[3333 rows x 6 columns]

In [52]:
pd.concat([left, right], axis=1)

State  Account length  Area code  Voice mail plan  Number vmail messages  \
0       KS             128        415             True                     25   
1       OH             107        415             True                     26   
2       NJ             137        415            False                      0   
3       OH              84        408            False                      0   
4       OK              75        415            False                      0   
...    ...             ...        ...              ...                    ...   
3328    AZ             192        415             True                     36   
3329    WV              68        415            False                      0   
3330    RI              28        510            False                      0   
3331    CT             184        510            False                      0   
3332    TN              74        415             True                     25   

      Total day minutes  Total day calls  Total day charge  Total eve minutes  
0                 265.1              110             45.07              197.4  
1                 161.6              123             27.47              195.5  
2                 243.4              114             41.38              121.2  
3                 299.4               71             50.90               61.9  
4                 166.7              113             28.34              148.3  
...                 ...              ...               ...                ...  
3328              156.2               77             26.55              215.5  
3329              231.1               57             39.29              153.4  
3330              180.8              109             30.74              288.8  
3331              213.8              105             36.35              159.6  
3332              234.4              113             39.85              265.9  

[3333 rows x 9 columns]

In [53]:
a = ["A", "B", "C"]
b = [1, 2, 3]

In [54]:
pd.DataFrame({"Letter": a, "Number": b})

Letter  Number
0      A       1
1      B       2
2      C       3